src/details/ArborX_DetailsDendrogram.hpp

```diff
@@ -0,0 +1,158 @@
+/****************************************************************************
+ * Copyright (c) 2017-2022 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef ARBORX_DETAILS_DENDROGRAM_HPP
+#define ARBORX_DETAILS_DENDROGRAM_HPP
+
+#include <ArborX_DetailsKokkosExtSwap.hpp>
+#include <ArborX_DetailsKokkosExtViewHelpers.hpp>
+#include <ArborX_DetailsSortUtils.hpp>
+#include <ArborX_DetailsUnionFind.hpp>
+#include <ArborX_MinimumSpanningTree.hpp> // WeightedEdge
+
+#include <Kokkos_Core.hpp>
+
+namespace ArborX::Details
+{
+
+// Compute permutation that orders edges in increasing order
+template <typename ExecutionSpace, typename MemorySpace>
+Kokkos::View<unsigned int *, MemorySpace>
+computeEdgesPermutation(ExecutionSpace const &exec_space,
+                        Kokkos::View<WeightedEdge *, MemorySpace> &edges)
+{
+  Kokkos::Profiling::pushRegion("ArborX::Dendrogram::dendrogram::sort_edges");
+
+  int const num_edges = edges.size();
+
+  Kokkos::View<float *, MemorySpace> weights(
+      Kokkos::view_alloc(exec_space, Kokkos::WithoutInitializing,
+                         "ArborX::Dendrogram::weights"),
+      num_edges);
+  Kokkos::parallel_for(
+      "ArborX::Dendrogram::copy_weights",
+      Kokkos::RangePolicy<ExecutionSpace>(exec_space, 0, num_edges),
+      KOKKOS_LAMBDA(int const e) { weights(e) = edges(e).weight; });
+
+  auto permute = Details::sortObjects(exec_space, weights);
+
+  Kokkos::Profiling::popRegion();
+
+  return permute;
+}
+
+template <typename Edges, typename UnionFind, typename Permute,
+          typename Parents>
+void dendrogramUnionFindHost(Edges edges_host, Permute permute,
+                             UnionFind union_find, Parents &parents_host)
+{
+  Kokkos::DefaultHostExecutionSpace exec_space;
```

Comment:
This is interesting.  If the execution space passed to the outer dendrogram calling function is on the host, we should probably use it.

Reply:
Only if it's `Serial` for the main loop. Could probably do the view initialization outside of the function using the provided execution space.